### PR TITLE
RLP-750: transport generalization

### PR DESCRIPTION
**Pull request by mortelli**
This PR adds the initial implementation for the **transport layer abstraction**, as described in the [Transport Research issue](https://jirainfuy.atlassian.net/browse/RLP-749). Presentations on this matter can be found [here](https://github.com/anarancio/lumino-docs/tree/master/presentations/transport).

This PR combines the result of multiple others (which build on this one) in order to carry out the implementation and solve the design problems previously laid out in the research documentation. The full list of PRs is [here](https://github.com/rsksmart/lumino/pulls?q=is%3Apr+label%3Atransport+is%3Aclosed).

~Tests are not being updated right now, so they will break. This will be tackled through a [separate issue](https://jirainfuy.atlassian.net/browse/RLP-775).~ Test suite is now functional; results can be found [here](https://github.com/anarancio/lumino-docs/tree/master/unit-testing/results/feature/RLP-750_transport-generalization).

After this PR is merged, the RIF Comms transport layer implementation can be added to the `rif-comms` branch. After this implementation is completed, `rif-comms` can be merged into `develop` and eventually `master`.

### CODE REVIEW

**Comment by mortelli** [comment]
although we don't plan on merging this right now, i am still going to request a preliminary review from @marcosmartinez7

**Review by JONAF2103 — COMMENTED** [review]
Looks good so far

File: raiden/network/transport/matrix/transport.py [hunk 1]
```diff
@@ -259,13 +251,14 @@ def __repr__(self):
...
 
-class MatrixTransport(Runnable):
+class MatrixTransport(TransportLayer, Runnable):
     _room_prefix = "raiden"
     _room_sep = "_"
     log = log
 
-    def __init__(self, config: dict, current_server_name: str = None):
-        super().__init__()
+    def __init__(self, address: Address, config: dict, current_server_name: str = None):
```
Comment (marcosmartinez7) [inline raiden/network/transport/matrix/transport.py:259]: Unused address param?

**Comment by mortelli** [comment]
25/09: the manual testing for full nodes running this branch was successful.

i've moved on to testing light clients, but it's turned out not to be a trivial process. 

i've decided to document the steps to create the manual testing environment, plus the concrete scenarios to try out.

**Comment by mortelli** [comment]
> 25/09: the manual testing for full nodes running this branch was successful.
> 
> i've moved on to testing light clients, but it's turned out not to be a trivial process.
> 
> i've decided to document the steps to create the manual testing environment, plus the concrete scenarios to try out.

documentation is now available on the [lumino-docs](https://github.com/anarancio/lumino-docs) repo.

i've manually verified basic operations in the following scenarios:
- `FN` ←→ `FN`
- `FN` ←→ `FN` ←→ `FN`
- `LC` ←→ `FN`
- `LC` ←→ `FN` ←→ `FN`

the following cases have been found to have issues which need to be fixed regardless of transport:
- `LC` ←→ `LC` (same or different hub)
- `LC` ←→ `FN` ←→ `LC` (same or different hub)

**Comment by mortelli** [comment]
we are now in the process of testing this branch, both manually and through the unit tests.

we've decided to merge once these pass.

**Comment by mortelli** [comment]
> we are now in the process of testing this branch, both manually and through the unit tests.
> 
> we've decided to merge once these pass.

unit tests have turned out OK (results [here](https://github.com/anarancio/lumino-docs/tree/master/unit-testing/results/feature/RLP-750_transport-generalization/2020-10-26)).

i am now in the process of running manual tests as described in [RLP-810](https://jirainfuy.atlassian.net/browse/RLP-810).

**Review by marcosmartinez7 — APPROVED** [review]

**Pull request merged**